wasm/src/language_service.rs

```diff
@@ -80,147 +81,142 @@ impl LanguageService {
-}
-
-#[derive(Serialize, Deserialize)]
-#[allow(non_snake_case)] // These types propagate to JS which expects camelCase
-pub struct CompletionItem {
-    pub label: String,
-    pub sortText: Option<String>,
-    pub kind: String,
-    pub detail: Option<String>,
-    pub additionalTextEdits: Option<Vec<TextEdit>>,
-}
-
-#[wasm_bindgen(typescript_custom_section)]
-const ITextEdit: &'static str = r#"
-export interface ITextEdit {
-    range: { start: number; end: number; };
-    newText: string;
+        additionalTextEdits?: ITextEdit[];
+    }"#
 }
-"#;
 
-#[derive(Serialize, Deserialize)]
-#[allow(non_snake_case)] // These types propagate to JS which expects camelCase
-pub struct TextEdit {
-    pub range: Span,
-    pub newText: String,
+serializable_type! {
+    struct TextEdit {
+        pub range: Span,
+        pub newText: String,
+    },
+    r#"export interface ITextEdit {
+        range: ISpan;
+        newText: string;
+    }"#
 }
 
-#[wasm_bindgen(typescript_custom_section)]
-const IHover: &'static str = r#"
-export interface IHover {
-    contents: string;
-    span: { start: number; end: number }
+serializable_type! {
+    struct Hover {
+        pub contents: String,
+        pub span: Span,
+    },
+    r#"export interface IHover {
+        contents: string;
+        span: ISpan
+    }"#,
+    Hover,
+    IHover,
+    "IHover"
 }
-"#;
 
-#[derive(Serialize, Deserialize)]
-pub struct Hover {
-    pub contents: String,
-    pub span: Span,
+serializable_type! {
+    struct Definition {
+        pub source: String,
+        pub offset: u32,
+    },
+    r#"export interface IDefinition {
+        source: string;
+        offset: number;
+    }"#,
+    Definition,
+    IDefinition,
+    "IDefinition"
 }
 
-#[wasm_bindgen(typescript_custom_section)]
-const IDefinition: &'static str = r#"
-export interface IDefinition {
-    source: string;
-    offset: number;
+serializable_type! {
+    struct Span {
+        pub start: u32,
+        pub end: u32,
+    },
+    r#"export interface ISpan {
+        start: number;
+        end: number;
+    }"#
 }
-"#;
 
-#[derive(Serialize, Deserialize)]
-pub struct Definition {
-    pub source: String,
-    pub offset: u32,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct Span {
-    pub start: u32,
-    pub end: u32,
-}
-
-#[wasm_bindgen(typescript_custom_section)]
-const IDiagnostic: &'static str = r#"
-export interface IDiagnostic {
-    start_pos: number;
-    end_pos: number;
-    message: string;
-    severity: "error" | "warning" | "info"
-    code?: {
-        value: string;
-        target: string;
-    }
+serializable_type! {
+    pub(crate) struct VSDiagnostic {
+        pub start_pos: usize,
+        pub end_pos: usize,
+        pub message: String,
+        pub severity: String,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub code: Option<VSDiagnosticCode>,
+    },
+    r#"export interface IDiagnostic {
+        start_pos: number;
```

Comment:
Some of the fields are snake case while others are camel case. I think the API actually wants camel case, but Rust wants snake case. Could we make them consistently all one way or the other? My preference is to do camel case as that is more true to the API, if it is not too much hassle to get Rust to not complain about it.

Reply:
I would also prefer all camel case, since these names inevitably propagate up to JS code. I actually added a warning suppression _inside_ the macro definition, so we should be able to use camel case freely for any types we define in the future.

I just... didn't want to touch this Diagnostic type for reasons I called out above... it's used in a bunch of places and I'm not super happy about how much we repeat ourselves. It would have made this PR even larger... so let's leave that cleanup for later.